app/controllers/application_controller.rb

```diff
@@ -12,12 +12,10 @@ class ApplicationController < ActionController::Base
 
   _helpers.module_eval { prepend DecoratorAdditions }
 
-  protect_from_forgery with: :reset_session # See ActionController::RequestForgeryProtection for details
+  protect_from_forgery with: ActionController::RequestForgeryProtection::ExceptionAndResetStrategy
 
   # Disable CSRF protection for requests to REST API.
-  skip_before_action :verify_authenticity_token, if: -> do
-    api_controller? && (params.key?(:provider_key) || params.key?(:access_token))
-  end
+  skip_forgery_protection if: -> { api_controller? && logged_in? }
```

Comment:
How does this work exactly?

```
  def logged_in?
    !!current_user
  end

  def current_user
    @current_user ||= (login_from_user_session || login_from_remember_me_cookie) unless @current_user == false
  end
```

If we have a user session, then we skip forgery protection? Doesn't sound right to me but maybe I'm not tracking down things deep enough.

Reply:
Yes, we skip forgery protection if we have a session, but **ONLY** for API requests. That's because API requests will never include a CSRF token, so we trust in the session.

Reply:
Do we allow session auth for API calls??

Either way, if there is no session, only token is specified (via query param or header), then why apply forgery protection? If yo udo a request with curl or seme API client, then you will not have a session, just token, no?

Reply:
> Do we allow session auth for API calls??

Sorry, no session, I chose the wrong term. I meant we only trust on authenticated API calls.

> Either way, if there is no session, only token is specified (via query param or header), then why apply forgery protection?

What this method does is precisely to skip forgery protection for authenticated API calls, since they will never include a CSRF token.

> If you do a request with curl or some API client, then you will not have a session, just token, no?

Right.

Reply:
This was my original point though, that the method seems to also return true when `login_from_user_session` is true. Without looking how it works, it seems like also accepting user sessions. And this seems to be wrong to me.

Reply:
Now I understand what you mean. API controllers override the `logged_in?` method so the one called is this:

https://github.com/3scale/porta/blob/f9ccd7442d8beb2e20634de64c1caaec252e05ff/app/lib/api_authentication/by_provider_key.rb#L9-L23

And the correct `current_user` is this one:

https://github.com/3scale/porta/blob/f9ccd7442d8beb2e20634de64c1caaec252e05ff/app/lib/api_authentication/by_access_token.rb#L7-L9

So it tries to get the user from the `access_token`, and in that fails, it tries to get the account from the `provider_token`. Cookies and sessions are not involved.

So IMO it shouldn't be a problem to keep it like this. My only concern is we have many controllers spread over a complex type hierarchy, so I wonder if there's a particular API controller which doesn't follow the flow I mentioned above. It's hard to say, but I can tell that everything inheriting from `Master::Api::BaseController` or `Admin::Api::BaseController` meets what I said above.

Reply:
Thinking about this a bit more... 
Why do we want to enable CSRF check on API controllers at all?... 

I think CSRF is really for the UI controllers, where the requests may take advantage of the stored session. On API controllers, a **valid** token is always required, otherwise a 403 will be thrown.

So, I think we don't need the extra CSRF check.

Moreover, the API controllers return a body for `403` depending on the requested format, e.g. 
```
<?xml version="1.0" encoding="UTF-8"?><error>Access denied</error>
```

```
{"error":"Access denied"}
```

While this is not extremely useful, I think it's nice. The "unverified token" handler, however, will not send any body.

So, I would actually consider disabling CSRF check for all API controllers.

Reply:
Agreed, if API controllers never use session auth, then it seems counter-productive to check CSRF token.

Reply:
OK, I disabled it for API endpoints.